File: flutter_app/lib/models/user_personalization.dart
class ListenerProfile {
  const ListenerProfile({
    required this.firstName,
    required this.country,
    required this.region,
    required this.city,
  });

  final String firstName;
  final String country;
  final String region;
  final String city;

  factory ListenerProfile.empty() {
    return const ListenerProfile(
      firstName: '',
      country: 'Romania',
      region: '',
      city: '',
    );
  }

  factory ListenerProfile.fromJson(Map<String, dynamic> json) {
    return ListenerProfile(
      firstName: json['first_name'] as String? ?? '',
      country: json['country'] as String? ?? 'Romania',
      region: json['region'] as String? ?? '',
      city: json['city'] as String? ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'first_name': firstName.trim(),
      'country': country.trim(),
      'region': region.trim(),
      'city': city.trim(),
    };
  }

  ListenerProfile copyWith({
    String? firstName,
    String? country,
    String? region,
    String? city,
  }) {
    return ListenerProfile(
      firstName: firstName ?? this.firstName,
      country: country ?? this.country,
      region: region ?? this.region,
      city: city ?? this.city,
    );
  }
}

class GeographyPreferenceMix {
  const GeographyPreferenceMix({
    required this.local,
    required this.national,
    required this.international,
  });

  final int local;
  final int national;
  final int international;

  factory GeographyPreferenceMix.defaults() {
    return const GeographyPreferenceMix(local: 34, national: 33, international: 33);
  }

  factory GeographyPreferenceMix.fromJson(Map<String, dynamic> json) {
    return GeographyPreferenceMix(
      local: _asInt(json['local'], 34),
      national: _asInt(json['national'], 33),
      international: _asInt(json['international'], 33),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'local': local,
      'national': national,
      'international': international,
    };
  }

  int get total => local + national + international;

  GeographyPreferenceMix copyWith({
    int? local,
    int? national,
    int? international,
  }) {
    return GeographyPreferenceMix(
      local: local ?? this.local,
      national: national ?? this.national,
      international: international ?? this.international,
    );
  }
}

class DomainPreferenceMix {
  const DomainPreferenceMix({
    required this.politics,
    required this.economy,
    required this.sport,
    required this.entertainment,
    required this.education,
    required this.health,
    required this.tech,
  });

  final int politics;
  final int economy;
  final int sport;
  final int entertainment;
  final int education;
  final int health;
  final int tech;

  factory DomainPreferenceMix.defaults() {
    return const DomainPreferenceMix(
      politics: 15,
      economy: 15,
      sport: 15,
      entertainment: 15,
      education: 10,
      health: 15,
      tech: 15,
    );
  }

  factory DomainPreferenceMix.fromJson(Map<String, dynamic> json) {
    return DomainPreferenceMix(
      politics: _asInt(json['politics'], 15),
      economy: _asInt(json['economy'], 15),
      sport: _asInt(json['sport'], 15),
      entertainment: _asInt(json['entertainment'], 15),
      education: _asInt(json['education'], 10),
      health: _asInt(json['health'], 15),
      tech: _asInt(json['tech'], 15),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'politics': politics,
      'economy': economy,
      'sport': sport,
      'entertainment': entertainment,
      'education': education,
      'health': health,
      'tech': tech,
    };
  }

  int get total =>
      politics + economy + sport + entertainment + education + health + tech;

  DomainPreferenceMix copyWith({
    int? politics,
    int? economy,
    int? sport,
    int? entertainment,
    int? education,
    int? health,
    int? tech,
  }) {
    return DomainPreferenceMix(
      politics: politics ?? this.politics,
      economy: economy ?? this.economy,
      sport: sport ?? this.sport,
      entertainment: entertainment ?? this.entertainment,
      education: education ?? this.education,
      health: health ?? this.health,
      tech: tech ?? this.tech,
    );
  }

  List<MapEntry<String, int>> rankedEntries() {
    final entries = <MapEntry<String, int>>[
      MapEntry('politics', politics),
      MapEntry('economy', economy),
      MapEntry('sport', sport),
      MapEntry('entertainment', entertainment),
      MapEntry('education', education),
      MapEntry('health', health),
      MapEntry('tech', tech),
    ];
    entries.sort((left, right) => right.value.compareTo(left.value));
    return entries;
  }
}

class EditorialPreferences {
  const EditorialPreferences({
    required this.geography,
    required this.domains,
  });

  final GeographyPreferenceMix geography;
  final DomainPreferenceMix domains;

  factory EditorialPreferences.defaults() {
    return EditorialPreferences(
      geography: GeographyPreferenceMix.defaults(),
      domains: DomainPreferenceMix.defaults(),
    );
  }

  factory EditorialPreferences.fromJson(Map<String, dynamic> json) {
    return EditorialPreferences(
      geography: GeographyPreferenceMix.fromJson(
        Map<String, dynamic>.from(json['geography'] as Map? ?? const {}),
      ),
      domains: DomainPreferenceMix.fromJson(
        Map<String, dynamic>.from(json['domains'] as Map? ?? const {}),
      ),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'geography': geography.toJson(),
      'domains': domains.toJson(),
    };
  }

  EditorialPreferences copyWith({
    GeographyPreferenceMix? geography,
    DomainPreferenceMix? domains,
  }) {
    return EditorialPreferences(
      geography: geography ?? this.geography,
      domains: domains ?? this.domains,
    );
  }
}

class UserPersonalization {
  const UserPersonalization({
    required this.listenerProfile,
    required this.editorialPreferences,
  });

  final ListenerProfile listenerProfile;
  final EditorialPreferences editorialPreferences;

  factory UserPersonalization.defaults() {
    return UserPersonalization(
      listenerProfile: ListenerProfile.empty(),
      editorialPreferences: EditorialPreferences.defaults(),
    );
  }

  factory UserPersonalization.fromJson(Map<String, dynamic> json) {
    return UserPersonalization(
      listenerProfile: ListenerProfile.fromJson(
        Map<String, dynamic>.from(json['listener_profile'] as Map? ?? const {}),
      ),
      editorialPreferences: EditorialPreferences.fromJson(
        Map<String, dynamic>.from(json['editorial_preferences'] as Map? ?? const {}),
      ),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'listener_profile': listenerProfile.toJson(),
      'editorial_preferences': editorialPreferences.toJson(),
    };
  }

  UserPersonalization copyWith({
    ListenerProfile? listenerProfile,
    EditorialPreferences? editorialPreferences,
  }) {
    return UserPersonalization(
      listenerProfile: listenerProfile ?? this.listenerProfile,
      editorialPreferences: editorialPreferences ?? this.editorialPreferences,
    );
  }

  String get locationSummary {
    final city = listenerProfile.city.trim();
    if (city.isEmpty) {
      return '${listenerProfile.country} / ${listenerProfile.region}';
    }
    return '${listenerProfile.country} / ${listenerProfile.region} / $city';
  }
}

int _asInt(Object? value, int fallback) {
  if (value is int) {
    return value;
  }
  if (value is double) {
    return value.round();
  }
  return fallback;
}

File: backend/app/models/user_personalization.py
from __future__ import annotations

from pydantic import BaseModel, Field, model_validator


_GEOGRAPHY_KEYS = ["local", "national", "international"]
_DOMAIN_KEYS = [
    "politics",
    "economy",
    "justice",
    "sport",
    "entertainment",
    "education",
    "health",
    "tech",
]


def _normalize_mix(values: dict[str, float], keys: list[str]) -> dict[str, float]:
    filtered = {key: max(float(values.get(key, 0.0)), 0.0) for key in keys}
    total = sum(filtered.values())

    if total <= 0:
        equal = round(100.0 / len(keys), 4)
        normalized = {key: equal for key in keys}
        normalized[keys[-1]] = round(100.0 - sum(normalized[key] for key in keys[:-1]), 4)
        return normalized

    if 0.99 <= total <= 1.01:
        scale = 100.0
    elif 99.0 <= total <= 101.0:
        scale = 1.0
    else:
        scale = 100.0 / total

    normalized = {key: round(value * scale, 4) for key, value in filtered.items()}
    normalized[keys[-1]] = round(100.0 - sum(normalized[key] for key in keys[:-1]), 4)
    return normalized


class ListenerProfile(BaseModel):
    first_name: str | None = None
    country: str | None = None
    region: str | None = None
    city: str | None = None

    def is_neutral(self) -> bool:
        return not any(
            value and str(value).strip()
            for value in [self.first_name, self.country, self.region, self.city]
        )

    def primary_local_anchor(self) -> str | None:
        region = (self.region or '').strip()
        if region:
            return region
        city = (self.city or '').strip()
        if city:
            return city
        return None

    def primary_local_anchor_scope(self) -> str:
        if (self.region or '').strip():
            return 'region'
        if (self.city or '').strip():
            return 'city_fallback'
        return 'none'

    def primary_local_anchor_explanation(self) -> str:
        scope = self.primary_local_anchor_scope()
        anchor = self.primary_local_anchor()
        if scope == 'region' and anchor:
            return f"Local editorial anchor uses listener region or county '{anchor}' as the primary local reference."
        if scope == 'city_fallback' and anchor:
            return f"Listener city '{anchor}' is stored and used only as a fallback because no region or county was provided."
        return 'No local editorial anchor was provided in the listener profile.'


class GeographyPreferenceMix(BaseModel):
    local: float = Field(default=0.0, ge=0.0, le=100.0)
    national: float = Field(default=0.0, ge=0.0, le=100.0)
    international: float = Field(default=0.0, ge=0.0, le=100.0)

    @model_validator(mode="after")
    def normalize(self) -> "GeographyPreferenceMix":
        normalized = _normalize_mix(self.model_dump(), _GEOGRAPHY_KEYS)
        for key, value in normalized.items():
            object.__setattr__(self, key, value)
        return self


class DomainPreferenceMix(BaseModel):
    politics: float = Field(default=0.0, ge=0.0, le=100.0)
    economy: float = Field(default=0.0, ge=0.0, le=100.0)
    justice: float = Field(default=0.0, ge=0.0, le=100.0)
    sport: float = Field(default=0.0, ge=0.0, le=100.0)
    entertainment: float = Field(default=0.0, ge=0.0, le=100.0)
    education: float = Field(default=0.0, ge=0.0, le=100.0)
    health: float = Field(default=0.0, ge=0.0, le=100.0)
    tech: float = Field(default=0.0, ge=0.0, le=100.0)

    @model_validator(mode="after")
    def normalize(self) -> "DomainPreferenceMix":
        normalized = _normalize_mix(self.model_dump(), _DOMAIN_KEYS)
        for key, value in normalized.items():
            object.__setattr__(self, key, value)
        return self


class EditorialPreferenceProfile(BaseModel):
    geography: GeographyPreferenceMix = Field(default_factory=GeographyPreferenceMix)
    domains: DomainPreferenceMix = Field(default_factory=DomainPreferenceMix)

    def is_neutral(self) -> bool:
        default_profile = EditorialPreferenceProfile()
        return self.model_dump() == default_profile.model_dump()


class UserPersonalization(BaseModel):
    listener_profile: ListenerProfile = Field(default_factory=ListenerProfile)
    editorial_preferences: EditorialPreferenceProfile = Field(default_factory=EditorialPreferenceProfile)

    def personalization_used(self) -> bool:
        return self.listener_profile_used() or self.editorial_preferences_used()

    def listener_profile_used(self) -> bool:
        return not self.listener_profile.is_neutral()

    def editorial_preferences_used(self) -> bool:
        return not self.editorial_preferences.is_neutral()

    def defaults_applied(self) -> bool:
        return not self.personalization_used()

    def local_editorial_anchor(self) -> str | None:
        return self.listener_profile.primary_local_anchor()

    def local_editorial_anchor_scope(self) -> str:
        return self.listener_profile.primary_local_anchor_scope()

    def local_editorial_anchor_explanation(self) -> str:
        return self.listener_profile.primary_local_anchor_explanation()

    def explainability(self) -> tuple[bool, bool, bool, bool, str]:
        personalization_used = self.personalization_used()
        listener_profile_used = self.listener_profile_used()
        editorial_preferences_used = self.editorial_preferences_used()
        defaults_applied = self.defaults_applied()
        local_anchor_note = self.local_editorial_anchor_explanation()
        explanation = (
            f"User personalization was provided and carried through the pipeline. {local_anchor_note}"
            if personalization_used
            else "Pipeline used safe neutral personalization defaults because no explicit personalization payload was provided."
        )
        return (
            personalization_used,
            listener_profile_used,
            editorial_preferences_used,
            defaults_applied,
            explanation,
        )

    @classmethod
    def from_input(
        cls,
        personalization: UserPersonalization | None = None,
        editorial_preferences: EditorialPreferenceProfile | None = None,
        listener_profile: ListenerProfile | None = None,
    ) -> "UserPersonalization":
        if personalization is not None:
            return personalization
        if editorial_preferences is not None or listener_profile is not None:
            return cls(
                listener_profile=listener_profile or ListenerProfile(),
                editorial_preferences=editorial_preferences or EditorialPreferenceProfile(),
            )
        return cls()
